### ISSUE TITLE
Type Incompatibility with Protocols Involving Dataclasses

### ISSUE THREAD

**Issue by ehupin** [microsoft/pyright#8196]
I am working with dataclasses and protocols, and I would expect the following code to pass type-checker, but it doesn't.
Am I doing something wrong, is this the expected behavior?

```python
from dataclasses import dataclass
from typing import Any, ClassVar, Dict, Protocol

# here is a Protocol that should match any dataclass
class DataclassProtocol(Protocol):
    __dataclass_fields__: ClassVar[Dict[str, Any]]

################ Experimenting with functions - works as expected ################

# here are two classes, only the first one is a dataclass
@dataclass
class ThisIsADataclass:
    a: int
    b: str

class ThisIsNotADataclass:
    def __init__(self, a: int, b: str) -> None:
        self.a = a
        self.b = b

# and here is a function that should accept any dataclass
def do_something_with_dataclass(obj: DataclassProtocol):
    ...

# type checker is happy with the first call, but not with the second, which is expected
do_something_with_dataclass(ThisIsADataclass(1, "a"))
do_something_with_dataclass(ThisIsNotADataclass(1, "a"))
# Argument of type "ThisIsNotADataclass" cannot be assigned to parameter "obj" of type "DataclassProtocol" in function "do_something_with_dataclass"
#   "ThisIsNotADataclass" is incompatible with protocol "DataclassProtocol"
#     "__dataclass_fields__" is not present

################ Experimenting with classes and methods - fails ################

# now let's create a Protocol for a class that requires a dataclass for one of its methods
class DataclassProcessorProtocol(Protocol):
    def do_something(self, obj: DataclassProtocol) -> None:
        ...

# and here is a class that should implement the protocol
class SpecificDataclassProcessor:
    def do_something(self, obj: ThisIsADataclass) -> None:
        ...

# and here is a function that should accept any class that implements the protocol
def use_dataclass_processor(processor: DataclassProcessorProtocol):
    ...

# but the type checker is not happy when I pass an instance of the class that I thought match the DataclassProcessorProtocol protocol 
use_dataclass_processor(SpecificDataclassProcessor())

# Argument of type "SpecificDataclassProcessor" cannot be assigned to parameter "processor" of type "DataclassProcessorProtocol" in function "use_dataclass_processor"
#   "SpecificDataclassProcessor" is incompatible with protocol "DataclassProcessorProtocol"
#     "do_something" is an incompatible type
#       Type "(obj: ThisIsADataclass) -> None" is incompatible with type "(obj: DataclassProtocol) -> None"
#         Parameter 1: type "DataclassProtocol" is incompatible with type "ThisIsADataclass"
#           "DataclassProtocol" is incompatible with "ThisIsADataclass"
```

**Comment by erictraut** [comment]
Pyright's behavior here is correct, so this isn't a bug.

The method `DataclassProcessorProtocol.do_something` accepts an argument that is type `DataclassProtocol`, but `SpecificDataclassProcessor.do_something` accepts an argument that is type `ThisIsADataclass`. This means `use_dataclass_processor` could call the `do_something` method and pass an object that doesn't derive from `ThisIsADataclass` but still matches the `DataclassProtocol` protocol.

This may seem unintuitive, but parameters of a callable are implicitly contravariant. You are attempting to use a covariant type here. `ThisIsADataclass` is a subtype of `DataclassProtocol`, but the converse is not true.

You can fix this in your code by changing `SpecificDataclassProcessor.do_something` to accept an argument of type `DataclassProtocol`.

**Comment by ehupin** [comment]
Thanks for the clarification :pray:
